File: Difficulty: Hard/Find median in a stream/find-median-in-a-stream.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    vector<double> getMedian(vector<int> &arr) {
        // code here
        vector<double> ans;
       priority_queue<int> maxHeap;
       priority_queue<int, vector<int>, greater<int>> minHeap;
       ans.push_back(arr[0]);
       minHeap.push(arr[0]);
        for(int i=1;i<arr.size();i++){
            int curr = arr[i];
            if(i % 2 != 0){
                int minInMinHeap = minHeap.top();
                minHeap.pop();
                minHeap.push(max(curr, minInMinHeap));
                maxHeap.push(min(curr, minInMinHeap));
                ans.push_back(((double)minHeap.top() + maxHeap.top())/2.0);
            }else{
                int minInMinHeap = minHeap.top();
                int maxInMaxHeap = maxHeap.top();
                if(curr > minInMinHeap){
                    minHeap.push(curr);
                    ans.push_back(minInMinHeap);
                }else if(curr < maxInMaxHeap){
                    maxHeap.pop();
                    minHeap.push(maxInMaxHeap);
                    maxHeap.push(curr);
                    ans.push_back(maxInMaxHeap);
                }else{
                    minHeap.push(curr);
                    ans.push_back(curr);
                }
            }
        }
        return ans;
    }
};


//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {

        string s;
        getline(cin, s);
        stringstream ss(s);
        vector<int> nums;
        int num;
        while (ss >> num) {
            nums.push_back(num);
        }
        Solution ob;
        vector<double> ans = ob.getMedian(nums);
        cout << fixed << setprecision(1);
        for (auto &i : ans)
            cout << i << " ";
        cout << "\n";
        cout << "~" << endl;
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Hard/Find median in a stream/find-median-in-a-stream.java
class Solution {
    void addNum(int num,PriorityQueue<Integer>minHeap,PriorityQueue<Integer>maxHeap){
        if(maxHeap.isEmpty()||num<=maxHeap.peek()){
            maxHeap.add(num);
        }
         else{
             minHeap.add(num);
         }
         if(maxHeap.size()>minHeap.size()+1){
             minHeap.add(maxHeap.poll());
         }
         else if(minHeap.size()>maxHeap.size()){
             maxHeap.add(minHeap.poll());
         }
    }
    double findMedian(PriorityQueue<Integer>minHeap,PriorityQueue<Integer>maxHeap){
        if(minHeap.size()==maxHeap.size()){
            return (minHeap.peek()+maxHeap.peek())/2.0;
        }
        return maxHeap.peek();
    }
    public ArrayList<Double> getMedian(int[] arr) {
        // code here
        PriorityQueue<Integer>minHeap=new PriorityQueue<>();
        PriorityQueue<Integer>maxHeap=new PriorityQueue<>((a,b)->b-a);
        ArrayList<Double>res=new ArrayList<>();
        for(int num:arr){
            addNum(num,minHeap,maxHeap);
            res.add(findMedian(minHeap,maxHeap));
        }
        return res;
        
    }
    
}